Add unit tests to FeeLineDetailsViewModelTests

thehenrybyrd opened the pull request:
### Description
This PR adds two assertions to test items that were not yet tested.
**First:** `.isExistingFeeLine` already has an `AssertTrue` in `test_view_model_prefills_input_data_correctly`, but an `AssertFalse` can alert us to a potential situation where initialization of `.isExistingFeeLine` is broken.
**Second:** `amountPlaceholder` was not being tested yet. It should always be `"0"`, so I used AssertEqual to ensure we do not pass an empty string instead of `"0"`.

### Testing instructions
Both asserts should pass if you run `test_view_model_formats_amount_correctly`.
Then, try to make them fail:
1. Try initializing with `.isExistingFeeLine` set to true. (line 116 of `FeeLineDetailsViewModel`)
2. Try passing an empty string, or any string other than "0". (line 114 of `FeeLineDetailsViewModel`)

Can you think of other ways to break these? Especially `.isExistingFeeLine`?

<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

One thing I am wondering:
I placed these within `test_view_model_formats_amount_correctly`, as it uses a blank initial viewModel. However, these asserts do not have to do with formatting the amount.
Would it be best to: create a new test, rename the existing test, or not worry about it since the asserts will make it clear what part failed?

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

thehenrybyrd committed:
Add AssertFalse for .isExistingFeeLine

This will alert us if the initializer's setup of .isExistingFeeLine breaks.

thehenrybyrd committed:
Add AssertEqual for .amountPlaceholder

Ensures that .amountPlaceholder does not have an empty string.

rachelmcr reviewed (approved):
Thanks for adding these assertions! The placeholder check looks especially helpful here. My only concern is the one you raised about where to put these:

> I placed these within test_view_model_formats_amount_correctly, as it uses a blank initial viewModel. However, these asserts do not have to do with formatting the amount.
Would it be best to: create a new test, rename the existing test, or not worry about it since the asserts will make it clear what part failed?

In this case, I'd suggest creating one or more separate tests, even if it uses the same view model setup. That way, if an assertion fails the test name can clearly reflect what caused the failure / what part of the view model the failure is related to.

thehenrybyrd committed:
Merge branch 'trunk' into add/fee-line-details-tests

thehenrybyrd committed:
Revert "Merge branch 'trunk' into add/fee-line-details-tests"

This reverts commit b5281618752c44a18ebd7e8788889fd719b348a3, reversing
changes made to 1f67c12ca41f59250fad4ce722a66c518da69418.

thehenrybyrd committed:
Move asserts into two new tests

test_view_model_amount_placeholder_is_not_empty_string()
and
test_view_model_initializes_correctly_with_no_existing_fee_line()

thehenrybyrd commented:
@rachelmcr thanks so much!
I've moved both asserts into new tests:

`test_view_model_amount_placeholder_is_not_empty_string()` does basic setup and then asserts that `amountPlaceholder` is equal to `"0"`. 
`test_view_model_initializes_correctly_with_no_existing_fee_line()` does the same setup, adds an amount, and then checks that `isPercentageOptionAvailable` is false and `isExistingFeeLine` is false.
I moved `isPercentageOptionAvailable` because it was in `test_view_model_formats_amount_correctly()` but doesn't really have to do with formatting the amount. As I understand, `isPercentageOptionAvailable` describes whether the "%" tab is displayed, and should be false whenever `baseAmountForPercentage` is not equal to 0.

Does that all sound reasonable? Do the names make sense?

Note: I merged trunk, but it resulted in an `Abort trap`, so I reverted the merge. Should have checked that trunk was building before I merged!

rachelmcr commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/FeeLineDetailsViewModelTests.swift:
```diff
@@ -209,4 +206,34 @@ final class FeeLineDetailsViewModelTests: XCTestCase {
         XCTAssertEqual(savedFeeLine?.total, "20.00")
     }
+
+    func test_view_model_amount_placeholder_is_not_empty_string() {
```
nit: You could use a name like `test_view_model_amount_placeholder_has_expected_value` since you're specifically checking that it is set to `"0"` (and it could, in theory, fail because the value is set to a different number rather than an empty string).

rachelmcr reviewed (changes requested):
The new tests look good! I made a small suggestion about the naming, but nothing major.

However, something has gone wrong with the merge/revert. It looks like the merge worked but since it was reverted, it's reverting all the changes that were in the merge. (You can see that the number of lines changes in the diff jumped up because of that.) That should be fixable by reverting the revert commit ([ce28086](https://github.com/woocommerce/woocommerce-ios/pull/6347/commits/ce280866532440e3a4bea0d072f0965c173e9518)) but let me know in Slack if you'd like to walk through that together.

thehenrybyrd commented:
I ended up in a big pile of Git trouble trying to rebase & reset, so I'm closing this PR and starting over with https://github.com/woocommerce/woocommerce-ios/pull/6428.
